### PR TITLE
MDEV-28926 Add time spent on query optimizer to JSON ANALYZE

### PR DESCRIPTION
**Pull request by mariadb-LuisLizardo**
- [x] *The Jira issue number for this PR is: MDEV-28926*

## Description
ANALYZE FORMAT=JSON does not show the time spent doing query optimization. This PR adds it with the following format: 

```
"query_optimization": {  "r_total_time_ms": NNNN.NNN }
```

The query optimization time is measured between the calls to `Explain_query::Explain_query()` and `Explain_query::query_plan_ready()`.

With this PR, the ANALYZE FORMAT=JSON output will become: 

```
MariaDB [j1]> show analyze format=json for 6;
--
{
   "r_query_time_in_progress_ms": 60499,
   "query_optimization": {
      "r_total_time_ms": "NNNN.NNN"
   },
   "query_block": {
      "select_id": 1,
      "r_loops": 1,
      "nested_loop": [
         {
            "table": {
               "table_name": "t1",
               "access_type": "ALL",
               "r_loops": 1,
               "rows": 1000,
               "r_rows": 1,
               "r_table_time_ms": 0.156016501,
               "r_other_time_ms": 0,
               "filtered": 100,
               "r_filtered": 100,
               "attached_condition": "t1.a <= 3"
            }
         }
      ]
   }
}
```

Besides that, this PR improves `Explain_query::print_explain_json()` quality by adding more modularity and verbose names.
Fix the problem of the writer being left open when no query plans were found.   

## How can this PR be tested?

Test cases at `mysql-test/` were adapted to include the new JSON element. These tests covered all changes. 

## Basing the PR against the correct MariaDB version
- [x] *This is a new feature and the PR is based against the latest MariaDB development branch*
- [ ] *This is a bug fix and the PR is based against the earliest branch in which the bug can be reproduced*

### CODE REVIEW

File: sql/sql_explain.h [hunk 1]
```diff
@@ -533,6 +537,8 @@ class Explain_query : public Sql_alloc
 #ifndef DBUG_OFF
   bool can_print_json= false;
 #endif
+
+  Exec_time_tracker query_optimization_total_time_tracker;
```
Comment (spetrunia) [inline sql/sql_explain.h:541]: This is a very long name. Please change for a shorter one, `optimization_time_tracker`.

File: sql/sql_explain.cc [hunk 2]
```diff
@@ -256,44 +259,75 @@ int Explain_query::print_explain_json(select_result_sink *output,
...
+    if (query_time_in_progress_ms > 0){
+      writer.add_member("r_query_time_in_progress_ms").
+            add_ull(query_time_in_progress_ms);
+    }
+
+    print_query_optimization_json(&writer);
+    is_show_cmd = true;
+  }
+
+  bool were_query_plans_found = print_query_blocks_json(&writer, is_analyze, is_show_cmd);
```
Comment (spetrunia) [inline sql/sql_explain.cc:279]: Also a very long name. Please change to `plan_found`  (it's singular. there can't be more than one).

File: sql/sql_explain.cc [hunk 3]
```diff
@@ -256,44 +259,75 @@ int Explain_query::print_explain_json(select_result_sink *output,
...
+  {
+    // if more timers are added, move the query_optimization member to outside the if
+    writer->add_member("query_optimization").start_object();
+    writer->add_member("r_total_time_ms").
+            add_double(query_optimization_total_time_tracker.get_time_ms());
+    writer->end_object(); 
+  }
+}
+
+bool Explain_query::print_query_blocks_json(Json_writer *writer, const bool is_analyze, const bool is_show_cmd)
```
Comment (spetrunia) [inline sql/sql_explain.cc:302]: Please keep the lines to be no longer than 80 characters.
(check the code around to see how multi-line code is formatted)

File: sql/sql_explain.cc [hunk 4]
```diff
@@ -256,44 +259,75 @@ int Explain_query::print_explain_json(select_result_sink *output,
...
-    node->print_explain_json(this, &writer, is_analyze, is_show_cmd);
+      return false; /* No query plan */
+    node->print_explain_json(this, writer, is_analyze, is_show_cmd);
   }
 
-  writer.end_object();
+  return true;
+}
 
+void Explain_query::send_explain_json_to_output(Json_writer *writer, select_result_sink *output)
```
Comment (spetrunia) [inline sql/sql_explain.cc:320]: The same as above: please make the lines to be no longer than 80 chars. max

File: sql/sql_explain.cc [hunk 5]
```diff
@@ -256,44 +259,75 @@ int Explain_query::print_explain_json(select_result_sink *output,
...
+
+    print_query_optimization_json(&writer);
+    is_show_cmd = true;
+  }
+
+  bool were_query_plans_found = print_query_blocks_json(&writer, is_analyze, is_show_cmd);
 
+  writer.end_object();
+
+  if(were_query_plans_found){
```
Comment (spetrunia) [inline sql/sql_explain.cc:283]: Coding style: please use space after `if` and put the `{` onto the next line:

```
 if ( )
 {
 ```

**Comment by spetrunia** [comment]
Generally, looks ok.  
There are a few fixes that need to be made:
* Please address the comments in the review.
* Please make all changes be in one patch.
Please push into branch named `bb-10.10-mdev28926`.  (the `bb-` part is important so that CI will pick it up).

**Review by spetrunia — CHANGES_REQUESTED** [review]
The changes needed in the patch

**Comment by mariadb-LuisLizardo** [comment]
@spetrunia what do you mean by one patch? One commit? 

I addressed all the change requests and push it, under the branch `bb-10.10-mdev28926` as you requested. 
For that, i had to open a new PR. https://github.com/MariaDB/server/pull/2193

**Pull request closed without merging**